workers/unity/Packages/io.improbable.worker.sdk/EventTracing.cs

```diff
@@ -0,0 +1,349 @@
+                    {
+                        newSpan.Causes[i] = new SpanId();
+                        fixed (byte* spanIdDest = newSpan.Causes[i].Data)
+                        {
+                            ApiInterop.Memcpy(spanIdDest, itemContainer->ItemUnion.Span.Causes[i].Data, (UIntPtr) 16);
+                        }
+                    }
+
+                    break;
+                case (byte) ItemType.Event:
+                    newItem.ItemType = ItemType.Event;
+                    newItem.TraceItem = new Event();
+                    var newEvent = (Event) newItem.TraceItem;
+
+                    newEvent.Id = new SpanId();
+                    ApiInterop.Memcpy(newEvent.Id.Data, itemContainer->ItemUnion.Event.Id.Data, (UIntPtr) 16);
+
+                    newEvent.UnixTimestampMillis = itemContainer->ItemUnion.Event.UnixTimestampMillis;
+                    newEvent.Type = ApiInterop.FromUtf8Cstr(itemContainer->ItemUnion.Event.Type);
+                    newEvent.Message = ApiInterop.FromUtf8Cstr(itemContainer->ItemUnion.Event.Message);
+
+                    newEvent.Data = Marshal.PtrToStructure<TraceEventData>(itemContainer->ItemUnion.Event.Data);
+                    break;
+                default:
+                    throw new NotSupportedException("Invalid Item Type provided.");
+            }
+
+            return newItem;
+        }
+
+        private CEventTrace.Item* ConvertItem(Item item)
+        {
+            var newItem = CEventTrace.ItemCreate(storage, null);
+            newItem->ItemUnion = new CEventTrace.Item.Union();
+
+            switch (item.ItemType)
+            {
+                case ItemType.Span:
+                    var spanItem = (Span) item.TraceItem;
+                    newItem->ItemUnion.Span = new CEventTrace.Span();
+                    newItem->ItemUnion.Span.Id = ParameterConversion.ConvertSpanId(spanItem.Id);
+                    newItem->ItemUnion.Span.CauseCount = (uint) spanItem.CauseCount;
+                    for (var i = 0; i < newItem->ItemUnion.Span.CauseCount; i++)
+                    {
+                        newItem->ItemUnion.Span.Causes[i] = ParameterConversion.ConvertSpanId(spanItem.Causes[i]);
+                    }
+
+                    break;
+                case ItemType.Event:
+                    var eventItem = (Event) item.TraceItem;
+                    newItem->ItemUnion.Event = ParameterConversion.ConvertEvent(eventItem);
+                    break;
+                default:
+                    throw new NotSupportedException("Invalid Item Type provided.");
+            }
+
+            newItem->ItemType = (byte) item.ItemType;
+
+            return newItem;
+        }
+    }
+
+    public unsafe delegate void TraceCallback(
+        UIntPtr userData, IntPtr item);
+
+    public class EventTracerParameters
+    {
+        public TraceCallback TraceCallback;
+        public UIntPtr UserData;
+    }
```

Comment:
These should use non-pointer types: 

```csharp
public delegate void TraceCallback(object userData, Item item);

public class EventTracerParameters 
{
	public TraceCallback Callback;
	public object UserData
}
```

and we should handle the interop in the library itself (so users don't need to worry about it).